components/object_store/src/mem_cache.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! An implementation of ObjectStore, which support
+//! 1. Cache based on memory, and support evict based on memory usage
+//! 2. Builtin Partition to reduce lock contention
+
+use std::{
+    collections::hash_map::DefaultHasher,
+    fmt::Display,
+    hash::{Hash, Hasher},
+    ops::Range,
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::stream::BoxStream;
+use lru_weighted_cache::{LruWeightedCache, Weighted};
+use tokio::{io::AsyncWrite, sync::Mutex};
+use upstream::{path::Path, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore, Result};
+
+struct CachedBytes(Bytes);
+
+impl Weighted for CachedBytes {
+    fn weight(&self) -> usize {
+        self.0.len()
+    }
+}
+
+#[derive(Debug)]
+struct Partition {
+    inner: Mutex<LruWeightedCache<String, CachedBytes>>,
+}
+
+impl Partition {
+    fn new(mem_cap: usize) -> Self {
+        Self {
+            inner: Mutex::new(LruWeightedCache::new(1, mem_cap).expect("invalid params")),
+        }
+    }
+}
+impl Partition {
+    async fn get(&self, key: &String) -> Option<Bytes> {
+        let mut guard = self.inner.lock().await;
+        guard.get(key).map(|v| v.0.clone())
+    }
+
+    async fn insert(&self, key: String, value: Bytes) {
+        let mut guard = self.inner.lock().await;
+        // don't care error now.
+        _ = guard.insert(key, CachedBytes(value));
+    }
+}
+
+#[derive(Debug)]
+struct MemCache {
+    /// Max memory this store can use
+    mem_cap: usize,
+    partitions: Vec<Arc<Partition>>,
+    partition_mask: usize,
+}
+
+impl MemCache {
+    fn new(partition_bits: usize, mem_cap: usize) -> Self {
+        let partition_num = 1 << partition_bits;
+        let cap_per_part = mem_cap / partition_num;
+        let partitions = (0..partition_num)
+            .map(|_| Arc::new(Partition::new(cap_per_part)))
+            .collect::<Vec<_>>();
+
+        Self {
+            mem_cap,
+            partitions,
+            partition_mask: partition_num - 1,
+        }
+    }
+
+    fn locate_partition(&self, key: &String) -> Arc<Partition> {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        self.partitions[hasher.finish() as usize & self.partition_mask].clone()
+    }
+
+    async fn get(&self, key: &String) -> Option<Bytes> {
+        let partition = self.locate_partition(key);
+        partition.get(key).await
+    }
+
+    async fn insert(&self, key: String, value: Bytes) {
+        let partition = self.locate_partition(&key);
+        partition.insert(key, value).await;
+    }
+}
+
+impl Display for MemCache {
```

Comment:
Why impl `Display` for it rather than `Debug`? 
Do we need to call `to_string` in somewhere? If just debug, maybe impl `Debug` is enough?